### PR TITLE
FIX: allow for print_level greater than 1 and ENH: allow warm start for dual variables

### LINKED ISSUE #234: `print_level` doesn't seem to work correctly for `minimize_ipopt`
When I give `options = {"disp": 5}` to `minimize_ipopt` it always ends up using `{b"print_level": 0}`, meaning no output is printed. I believe the problem is [here in the code](https://github.com/mechmotum/cyipopt/blob/d00088053569a6486510ed8c2baafb38ad3eb6a2/cyipopt/scipy_interface.py#L585C1-L588C36). Given that `options` is a `dict` it will always end up in the `else` case. Maybe `option.get()` is more suitable here? And it would be great if we can allow for display levels greater than `1`. Apologies if I'm missing something or using the interface wrong!

### PR DESCRIPTION
**Pull request by zhengzl18**
fix #234 and #151

### CODE REVIEW

File: cyipopt/scipy_interface.py [hunk 1]
```diff
@@ -373,6 +373,9 @@ def _wrap_funs(fun, jac, hess, hessp, constraints, kwargs):
 
 def minimize_ipopt(fun,
                    x0,
+                   mult_g=[],
+                   mult_x_L=[],
+                   mult_x_U=[],
```
Comment (moorepants) [inline cyipopt/scipy_interface.py:378]: Please add new kwargs to the end of this list of kwargs to maintain backwards compatibility of order.

**Comment by moorepants** [comment]
Can you please add unit tests for these changes? Thanks for the contribution.

File: cyipopt/scipy_interface.py (hunk 1 above)
Comment (zhengzl18) [reply cyipopt/scipy_interface.py:378]: you mean like this?
```python
def minimize_ipopt(fun,
                   x0,
                   args=(),
                   kwargs=None,
                   method=None,
                   jac=None,
                   hess=None,
                   hessp=None,
                   bounds=None,
                   constraints=(),
                   tol=None,
                   callback=None,
                   options=None,
                   mult_g=[],
                   mult_x_L=[],
                   mult_x_U=[],):
```

(p.s. I haven't participated in such a big project before, and hence sorry if I introduced any error)

Comment (moorepants) [reply cyipopt/scipy_interface.py:378]: Yes, that's what I mean.

**Comment by zhengzl18** [comment]
> Can you please add unit tests for these changes? Thanks for the contribution.

Sorry, I don't quite know how to add unit test. I just tested with my own codes. Should I add some test script or something?

**Comment by moorepants** [comment]
We try to add a unit test, for example see https://github.com/mechmotum/cyipopt/blob/master/cyipopt/tests/unit/test_ipopt_funcs.py, to test any new additions/changes.

File: cyipopt/scipy_interface.py [hunk 2]
```diff
@@ -596,8 +609,11 @@ def minimize_ipopt(fun,
     # Rename some default scipy options
     replace_option(options, b'disp', b'print_level')
     replace_option(options, b'maxiter', b'max_iter')
-    if getattr(options, 'print_level', False) is True:
-        options[b'print_level'] = 1
+    if b'print_level' in options:
+        if options[b'print_level'] is True:
+            options[b'print_level'] = 1
+        elif options[b'print_level'] is False:
+            options[b'print_level'] = 0
```
Comment (brocksam) [inline cyipopt/scipy_interface.py:616]: This change doesn't appear to make any difference to the way the code functions, except increasing the number of lines of code to do the same thing. Am I missing something? If so, please can you explain why this change is needed?

Comment (brocksam) [reply cyipopt/scipy_interface.py:616]: Ipopt print levels are integers between 1 and 12, with the default being 5: https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_print_level

Comment (brocksam) [reply cyipopt/scipy_interface.py:616]: I think we should use something like the following. This will ensure that if a `print_level` of `True` is used that the default print level is used, not breaking previous code. It will also allow users to manually select the exact desired print level as per the Ipopt options.

```suggestion
    if getattr(options, 'print_level') is True:
        options[b'print_level'] = 5
    else:
        print_level = getattr(options, 'print_level' , 5)
        if not isinstance(print_level, int) or (print_level < 1) or (print_level > 12):
            raise ValueError(f'`print_level` option must be an int between 1 and 12, not {print_level}')
        options[b'print_level'] = print_level
            
```

File: cyipopt/scipy_interface.py [hunk 3]
```diff
@@ -385,7 +385,10 @@ def minimize_ipopt(fun,
                    constraints=(),
                    tol=None,
                    callback=None,
-                   options=None):
+                   options=None,
+                   mult_g=[],
```
[the remark this answers is not in the document]
Comment (zhengzl18) [reply cyipopt/scipy_interface.py:389]: Thanks for your patience! I did neglect this.

File: cyipopt/scipy_interface.py (hunk 2 above)
Comment (zhengzl18) [reply cyipopt/scipy_interface.py:616]: I thought it this way: The original version would force the `print_level` to be one if the user specify this parameter, no matter what value is chosen. I tried to make it work like this: 
1. 'disp': True -> 'print_level': 1
2. 'disp': False -> 'print_level': 0
3. no 'disp' or 'print_level' -> 'print_level': 0 (according to Ipopt's document, a `print_level` of value `0` is acceptable)
4. 'disp'/'print_level': 0-12 -> 'print_level': 0-12

About the default `print_level` in the second case, I did have some doubts. I wondered before why Ipopt said that `print_level` was by default `5`, but that was not how it actually behavioured. So when I was changing these codes, I hesitated about whether to set the default value to `1` or `5`. I chose `1` eventually because it was the original setting.

**Comment by zhengzl18** [comment]
That's weird. The checks seemed to fail because of the missing of SciPy. I don't understand what's going on. I've run `$ pytest` and passed all checks locally.

**Comment by moorepants** [comment]
Please merge in the tip of master to get the latest fixes for some scipy tests.

**Comment by zhengzl18** [comment]
> Please merge in the tip of master to get the latest fixes for some scipy tests.

It seems that all the tests I added in `test_dual_warm_start.py` should be skipped because they were all built upon `minimize_ipopt()` which involves SciPy. This somehow makes these tests meaningless. Am I right?

**Comment by moorepants** [comment]
We use the pytest skip decorator to skip the tests if scipy is not installed:

https://github.com/mechmotum/cyipopt/blob/master/cyipopt/tests/unit/test_scipy_optional.py#L23C2-L24C69

**Comment by zhengzl18** [comment]
Yeah I know. I mean, since all the tests I added are based on SciPy, skipping the scipy tests would make all these tests meaningless.

**Comment by zhengzl18** [comment]
Got it. Thanks for explanation.

File: cyipopt/scipy_interface.py [hunk 4]
```diff
@@ -691,3 +714,13 @@ def _minimize_ipopt_iv(fun, x0, args, kwargs, method, jac, hess, hessp,
 
     return (fun, x0, args, kwargs, method, jac, hess, hessp,
             bounds, constraints, tol, callback, options)
+
+def _dual_initial_guess_validation(dual_name: str, dual_value: list, length: int):
+    if not isinstance(dual_value, list):
+        raise TypeError(f'`{dual_name}` must be a list.')
```
Comment (moorepants) [inline cyipopt/scipy_interface.py:720]: Why does this have to be a list? Couldn't any iterable or sequence work?

File: cyipopt/scipy_interface.py [hunk 5]
```diff
@@ -385,7 +385,10 @@ def minimize_ipopt(fun,
                    constraints=(),
                    tol=None,
                    callback=None,
-                   options=None):
+                   options=None,
+                   mult_g=None,
+                   mult_x_L=None,
+                   mult_x_U=None,):
```
Comment (moorepants) [inline cyipopt/scipy_interface.py:391]: This function is supposed to mirror the scipy `minimize()` function. The new arguments you have added are not available in SciPy's `minimize()` function. Shouldn't then these new arguments be passed in through the `options` keyword which allows you to provide options for a specific solver?

File: cyipopt/scipy_interface.py (hunk 4 above)
Comment (zhengzl18) [reply cyipopt/scipy_interface.py:720]: Since `Problem.solve()` uses empty lists as the default values for `lagrange`, `zl` and `zu`, and create zero numpy arrays when they are `[]`, I just followed it. Indeed any iterable or sequence would work.
https://github.com/mechmotum/cyipopt/blob/20488262bae5647f6af0c176007631cbc687ec9d/cyipopt/cython/ipopt_wrapper.pyx#L628

File: cyipopt/scipy_interface.py (hunk 5 above)
Comment (zhengzl18) [reply cyipopt/scipy_interface.py:391]: I see. But I thought all keys in `options` are or could be transferred into options supported by ipopt (am I wrong?), wouldn't it be missleading if these three arguments are passed through `options`?
